### PR TITLE
Feature crud update

### PR DESCRIPTION
**Pull request by Sarahtonein**

### CODE REVIEW

File: controllers/rest/player/UpdatePlayer.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    	return c.JSON(http.StatusNotFound, "Failed to retrieve player from the database")
+	}
+
+	// Parse the JSON request body into a separate variable
+	updateData := struct {
+		Layer string `json:"layer"`
+		Position models.Position `json:"position"`
+		Rotation models.Rotation `json:"rotation"`
+		Scale    models.Scale    `json:"scale"`
+	}{}
```
Comment (Balugrizzly) [inline controllers/rest/player/UpdatePlayer.go:44]: Why are you not using the public model here?

**Review by Balugrizzly — COMMENTED** [review]
What is your plan with the update process 
specific to the data handling and binding?

**Comment by Balugrizzly** [comment]
But apart from the above section looks good

File: controllers/rest/player/UpdatePlayer.go (hunk 1 above)
Comment (Sarahtonein) [reply controllers/rest/player/UpdatePlayer.go]: > Why are you not using the public model here?

I was having a problem with using the model so opted for a new struct. I have modified this now.

File: README.md [hunk 2]
```diff
@@ -379,6 +380,98 @@ http://127.0.0.1:1323/player?active=true
 ]
 ```
 
+##### UpdatePlayer
+`PUT /player`
+
+Handles Updating Player Model Data. If no data is sent in JSON format then no data will be updated and saved into the database for a player.
+
```
Comment (Balugrizzly) [inline README.md:387]: Why are you using "PUT" 

The below sentence is in direct contradiction with the HTTP method

File: README.md [hunk 3]
```diff
@@ -379,6 +380,98 @@ http://127.0.0.1:1323/player?active=true
...
+**Query Parameters:**
+
+Name   | Type | Mandatory | 
+-------|------|-----------|
+ID     | INT |    YES     |
+
+
+**Request URL Example With all Accepted Query Params:**
+```html
+http://127.0.0.1:1323/1
```
Comment (Balugrizzly) [inline README.md:401]: Wrong URL and path?

File: README.md [hunk 4]
```diff
@@ -379,6 +380,98 @@ http://127.0.0.1:1323/player?active=true
...
+			"z": 6,
+			"w": 9
+		},
+		"position": {
+			"x": 1,
+			"y": 2,
+			"z": 6
+		}
+	}
+}
```
Comment (Balugrizzly) [inline README.md:419]: Invalid/Incomplete Json?

File: controllers/rest/player/UpdatePlayer.go
Comment (Balugrizzly) [inline controllers/rest/player/UpdatePlayer.go:1]: optional:
errors and error codes  are not documented

**Review by Balugrizzly — CHANGES_REQUESTED** [review]

**Comment by Balugrizzly** [comment]
Can you fix Merge conflicts?

File: controllers/rest/player/UpdatePlayer.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	updateData := models.Player{}
+	if err := c.Bind(&updateData); err != nil {
+		c.Logger().Error("Failed to parse update data")
+		return c.JSON(http.StatusBadRequest, "Invalid update data")
+	}
+
+	// Update the specific fields in queryPlayer
+	queryPlayer.Transform.Position = updateData.Transform.Position
+	queryPlayer.Transform.Rotation = updateData.Transform.Rotation
+	queryPlayer.Transform.Scale = updateData.Transform.Scale
```
[the remark this answers is not in the document]
Comment (Sarahtonein) [reply controllers/rest/player/UpdatePlayer.go:50]: > Why are you only updating those three selected fields and no others? Also does that work without handling the associated ID's?

Will add the other fields. And yes able to update the position and rotation for example without providing an ID for the position and rotation, please see below example:

image

File: README.md (hunk 2 above)
Comment (Sarahtonein) [reply README.md:387]: > Why are you using "PUT"
> 
> The below sentence is in direct contradiction with the HTTP method

Considering that PUT 'resets' unspecified fields I suppose we should be using PATCH HTTP method instead?

File: controllers/rest/player/UpdatePlayer.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Balugrizzly) [reply controllers/rest/player/UpdatePlayer.go]: Okey

[the remark this answers is not in the document]
Comment (Balugrizzly) [reply controllers/rest/player/UpdatePlayer.go]: let me know when you have added all the fields.
I will then resolve this convo

File: README.md (hunk 2 above)
Comment (Balugrizzly) [reply README.md]: Yes 
Will assume you made/make the necessary changes and makr this as resolved

File: README.md
Comment (Balugrizzly) [inline README.md:1]: Ähm why did you remove the entire Updateplayer section?

File: controllers/rest/player/UpdatePlayer.go [hunk 6]
```diff
@@ -45,10 +44,21 @@ func UpdatePlayer(c echo.Context) error {
 	}
 
 	// Update the specific fields in queryPlayer
+	queryPlayer.Layer = updateData.Layer
+	queryPlayer.ENS = updateData.ENS
+	queryPlayer.Active = updateData.Active
```
Comment (Balugrizzly) [inline controllers/rest/player/UpdatePlayer.go:49]: I highly suspect updating the active field wont work.
Look into updating false and null values

File: controllers/rest/player/UpdatePlayer.go [hunk 7]
```diff
@@ -45,10 +44,21 @@ func UpdatePlayer(c echo.Context) error {
...
 	queryPlayer.Transform.Scale = updateData.Transform.Scale
 
+	/*queryPlayer.States = []models.State{}
+
+	for _, state := range updateData.States {
+		queryPlayer.States = append(queryPlayer.States, models.State{
+			State: state.State,
+		})
+	}*/ 
+
```
Comment (Balugrizzly) [inline controllers/rest/player/UpdatePlayer.go:61]: What is this?

**Review by Balugrizzly — CHANGES_REQUESTED** [review]

File: controllers/rest/player/UpdatePlayer.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Sarahtonein) [reply controllers/rest/player/UpdatePlayer.go]: > let me know when you have added all the fields. I will then resolve this convo

Other fields have been added.

**Comment by Balugrizzly** [comment]
Are you planing to discuss this in Voice?
Asking as there are many open points after your revision

**Comment by Sarahtonein** [comment]
> Are you planing to discuss this in Voice? Asking as there are many open points after your revision

Yes, wanting to go over this in voice please.

**Comment by Balugrizzly** [comment]
Alright

File: controllers/rest/player/UpdatePlayer.go
[the remark this answers is not in the document]
Comment (Sarahtonein) [reply controllers/rest/player/UpdatePlayer.go:1]: GoFMT setup :)

File: controllers/rest/player/UpdatePlayer.go (hunk 6 above)
Comment (Sarahtonein) [reply controllers/rest/player/UpdatePlayer.go:50]: > I highly suspect updating the active field wont work. Look into updating false and null values

Updating active field like this does work, I had a look into updating false and null values and it took me down the discussed pointer hole.

File: controllers/rest/player/UpdatePlayer.go
[the remark this answers is not in the document]
Comment (Sarahtonein) [reply controllers/rest/player/UpdatePlayer.go:1]: > Optional
> 
> Support for Name and UserId

Added and will be modifying here depending on how this is structured in DELETE requests.

File: controllers/rest/player/UpdatePlayer.go (hunk 6 above)
Comment (Balugrizzly) [reply controllers/rest/player/UpdatePlayer.go:47]: Are you 100% sure? Have you tested this?
Do you wanna bet? :)

Comment (Sarahtonein) [reply controllers/rest/player/UpdatePlayer.go:47]: I did the following:

1. Created a player
```
{
	"ID": 19,
	"CreatedAt": "2023-10-30T23:59:14.416803326Z",
	"UpdatedAt": "2023-10-30T23:59:14.416803326Z",
	"DeletedAt": null,
	"UserID": "25f90f11-0306-462b-8781-04d8b47234e8",
	"name": "test_player",
	"layer": "",
	"ens": "",
	"active": true,
	"transform": {
		"RotationID": 83,
		"ScaleID": 83,
		"PositionID": 83,
		"position": {
			"ID": 83,
			"CreatedAt": "2023-10-30T23:59:14.40569386Z",
			"UpdatedAt": "2023-10-30T23:59:14.40569386Z",
			"DeletedAt": null,
			"x": 1,
			"y": 2,
			"z": 3
		},
		"rotation": {
			"ID": 83,
			"CreatedAt": "2023-10-30T23:59:14.411727103Z",
			"UpdatedAt": "2023-10-30T23:59:14.411727103Z",
			"DeletedAt": null,
			"x": 4,
			"y": 5,
			"z": 6,
			"w": 0
		},
		"scale": {
			"ID": 83,
			"CreatedAt": "2023-10-30T23:59:14.414415062Z",
			"UpdatedAt": "2023-10-30T23:59:14.414415062Z",
			"DeletedAt": null,
			"x": 7,
			"y": 8,
			"z": 9
		}
	},
	"states": null,
	"level": {
		"ID": 0,
		"CreatedAt": "0001-01-01T00:00:00Z",
		"UpdatedAt": "0001-01-01T00:00:00Z",
		"DeletedAt": null,
		"PlayerID": 0,
		"levelID": 0
	}
}
```

2. Set active to false
request body
```
{
  "UserID": "25f90f11-0306-462b-8781-04d8b47234e8",
  "Name": "test_player",
  "Layer": "layer1",
  "Ens": "",
	"Active": false,
  "Transform": {
    "Position": {
      "x": 1,
      "y": 2,
      "z": 3
    },
    "Rotation": {
      "x": 4,
      "y": 5,
      "z": 6
    },
    "Scale": {
      "x": 7,
      "y": 8,
      "z": 9
    }
  },
  "states": [
    {
      "id": 1,
      "value": 0.4
    },
    {
      "id": 2,
      "value": 0.1
    }
  ]
}
```

Response, and debug log from player service:

```
{
	"ID": 19,
	"CreatedAt": "2023-10-30T23:59:14.416803326Z",
	"UpdatedAt": "2023-10-31T00:01:27.077767629Z",
	"DeletedAt": null,
	"UserID": "25f90f11-0306-462b-8781-04d8b47234e8",
	"name": "test_player",
	"layer": "layer1",
	"ens": "",
	"active": false,
	"transform": {
		"RotationID": 84,
		"ScaleID": 84,
		"PositionID": 84,
		"position": {
			"ID": 84,
			"CreatedAt": "2023-10-31T00:01:27.070353261Z",
			"UpdatedAt": "2023-10-31T00:01:27.070353261Z",
			"DeletedAt": null,
			"x": 1,
			"y": 2,
			"z": 3
		},
		"rotation": {
			"ID": 84,
			"CreatedAt": "2023-10-31T00:01:27.073501709Z",
			"UpdatedAt": "2023-10-31T00:01:27.073501709Z",
			"DeletedAt": null,
			"x": 4,
			"y": 5,
			"z": 6,
			"w": 0
		},
		"scale": {
			"ID": 84,
			"CreatedAt": "2023-10-31T00:01:27.075746786Z",
			"UpdatedAt": "2023-10-31T00:01:27.075746786Z",
			"DeletedAt": null,
			"x": 7,
			"y": 8,
			"z": 9
		}
	},
	"states": [
		{
			"ID": 89,
			"CreatedAt": "2023-10-31T00:01:27.081366954Z",
			"UpdatedAt": "2023-10-31T00:01:27.081366954Z",
			"DeletedAt": null,
			"PlayerID": 19,
			"stateID": 0,
			"value": 0.4
		},
		{
			"ID": 90,
			"CreatedAt": "2023-10-31T00:01:27.081366954Z",
			"UpdatedAt": "2023-10-31T00:01:27.081366954Z",
			"DeletedAt": null,
			"PlayerID": 19,
			"stateID": 0,
			"value": 0.1
		}
	],
	"level": {
		"ID": 0,
		"CreatedAt": "0001-01-01T00:00:00Z",
		"UpdatedAt": "0001-01-01T00:00:00Z",
		"DeletedAt": null,
		"PlayerID": 0,
		"levelID": 0
	}
}
```
``
playerservice-playerservice-1  | {"time":"2023-10-31T00:01:27.070014479Z","level":"DEBUG","prefix":"echo","file":"UpdatePlayer.go","line":"47","message":"Player Active status before update:true"}
```` 
3. Stop player service

4. Start player service again

5. Set player to active

request body
```
{
  "UserID": "25f90f11-0306-462b-8781-04d8b47234e8",
  "Name": "test_player",
  "Layer": "layer1",
  "Ens": "",
	"Active": true,
  "Transform": {
    "Position": {
      "x": 1,
      "y": 2,
      "z": 3
    },
    "Rotation": {
      "x": 4,
      "y": 5,
      "z": 6
    },
    "Scale": {
      "x": 7,
      "y": 8,
      "z": 9
    }
  },
  "states": [
    {
      "id": 1,
      "value": 0.4
    },
    {
      "id": 2,
      "value": 0.1
    }
  ]
}

```
Response, and debug log from player service:

```
playerservice-playerservice-1  | {"time":"2023-10-31T00:02:29.567010124Z","level":"DEBUG","prefix":"echo","file":"UpdatePlayer.go","line":"47","message":"Player Active status before update:false"}
```

```
{
	"ID": 19,
	"CreatedAt": "2023-10-30T23:59:14.416803326Z",
	"UpdatedAt": "2023-10-31T00:02:29.575995316Z",
	"DeletedAt": null,
	"UserID": "25f90f11-0306-462b-8781-04d8b47234e8",
	"name": "test_player",
	"layer": "layer1",
	"ens": "",
	"active": true,
	"transform": {
		"RotationID": 85,
		"ScaleID": 85,
		"PositionID": 85,
		"position": {
			"ID": 85,
			"CreatedAt": "2023-10-31T00:02:29.567668357Z",
			"UpdatedAt": "2023-10-31T00:02:29.567668357Z",
			"DeletedAt": null,
			"x": 1,
			"y": 2,
			"z": 3
		},
		"rotation": {
			"ID": 85,
			"CreatedAt": "2023-10-31T00:02:29.570841487Z",
			"UpdatedAt": "2023-10-31T00:02:29.570841487Z",
			"DeletedAt": null,
			"x": 4,
			"y": 5,
			"z": 6,
			"w": 0
		},
		"scale": {
			"ID": 85,
			"CreatedAt": "2023-10-31T00:02:29.573016994Z",
			"UpdatedAt": "2023-10-31T00:02:29.573016994Z",
			"DeletedAt": null,
			"x": 7,
			"y": 8,
			"z": 9
		}
	},
	"states": [
		{
			"ID": 91,
			"CreatedAt": "2023-10-31T00:02:29.580191435Z",
			"UpdatedAt": "2023-10-31T00:02:29.580191435Z",
			"DeletedAt": null,
			"PlayerID": 19,
			"stateID": 0,
			"value": 0.4
		},
		{
			"ID": 92,
			"CreatedAt": "2023-10-31T00:02:29.580191435Z",
			"UpdatedAt": "2023-10-31T00:02:29.580191435Z",
			"DeletedAt": null,
			"PlayerID": 19,
			"stateID": 0,
			"value": 0.1
		}
	],
	"level": {
		"ID": 0,
		"CreatedAt": "0001-01-01T00:00:00Z",
		"UpdatedAt": "0001-01-01T00:00:00Z",
		"DeletedAt": null,
		"PlayerID": 0,
		"levelID": 0
	}
}
```

I think this covers the scenario in which you described but let me know if I''m off here. I have tested this the other way around too fwiw.

As mentioned if Active is not included in the request body it defaults to false.

Comment (Balugrizzly) [reply controllers/rest/player/UpdatePlayer.go:48]: okey i believe you and that you are right 
we want the active field to be omitable

File: CHANGELOG.md
Comment (Balugrizzly) [inline CHANGELOG.md:1]: This is wrong.
There is no 0.3 release

File: README.md [hunk 8]
```diff
@@ -387,6 +387,51 @@ http://127.0.0.1:1323/player?active=true
...
+Updates a player in the database. Takes a single parameter, identifiers
+that can be used include: UUID, ID, and Name.
+
+**Headers:** None
+
+**Path Parameters:**
+
+| Name | Type | Mandatory |
+| ---- | ---- | --------- |
+| ID   | INT  | YES       |
```
Comment (Balugrizzly) [inline README.md:403]: Type is inaccurate

**Review by Balugrizzly — COMMENTED** [review]
This looks good apart from the tiny two comments i made.
Where are we here in generall?

File: README.md (hunk 8 above)
Comment (Sarahtonein) [reply README.md]: Amended, thanks for catching that

File: CHANGELOG.md
Comment (Sarahtonein) [reply CHANGELOG.md:1]: removed 0.3

**Comment by Sarahtonein** [comment]
> This looks good apart from the tiny two comments i made. Where are we here in generall?

I think this is ready to merge from my tests. 

Example below:

URL: http://localhost:1323/player/2

JSON sent to service

```
{
  "UserID": "d1813663-cf15-4f33-a0c2-2d3b1c2fead5",
  "Name": "2",
  "Layer": "layer1",
  "Ens": "",
	"active": true,
  "Transform": {
    "Position": {
      "x": 4,
      "y": 5,
      "z": 6
    },
    "Rotation": {
      "x": 7,
      "y":	75,
      "z": 6
    },
    "Scale": {
      "x": 6,
      "y": 7,
      "z": 10
    }
  },
  "states": [
    {
      "stateid": 4,
      "value": 0.69
    },
    {
      "stateid": 20,
      "value": 0.420
    }
  ]
}
```

response from service:

```
{
	"ID": 2,
	"CreatedAt": "2023-11-05T21:56:28.17432539Z",
	"UpdatedAt": "2023-11-05T21:59:50.03589367Z",
	"DeletedAt": null,
	"UserID": "d1813663-cf15-4f33-a0c2-2d3b1c2fead5",
	"name": "2",
	"layer": "layer1",
	"ens": "",
	"active": true,
	"transform": {
		"RotationID": 175,
		"ScaleID": 175,
		"PositionID": 175,
		"position": {
			"ID": 175,
			"CreatedAt": "2023-11-05T21:59:49.962610279Z",
			"UpdatedAt": "2023-11-05T21:59:49.962610279Z",
			"DeletedAt": null,
			"x": 4,
			"y": 5,
			"z": 6
		},
		"rotation": {
			"ID": 175,
			"CreatedAt": "2023-11-05T21:59:50.000698312Z",
			"UpdatedAt": "2023-11-05T21:59:50.000698312Z",
			"DeletedAt": null,
			"x": 7,
			"y": 75,
			"z": 6,
			"w": 0
		},
		"scale": {
			"ID": 175,
			"CreatedAt": "2023-11-05T21:59:50.024145978Z",
			"UpdatedAt": "2023-11-05T21:59:50.024145978Z",
			"DeletedAt": null,
			"x": 6,
			"y": 7,
			"z": 10
		}
	},
	"states": [
		{
			"ID": 9,
			"CreatedAt": "2023-11-05T21:59:50.080820909Z",
			"UpdatedAt": "2023-11-05T21:59:50.080820909Z",
			"DeletedAt": null,
			"PlayerID": 2,
			"stateID": 4,
			"value": 0.69
		},
		{
			"ID": 10,
			"CreatedAt": "2023-11-05T21:59:50.080820909Z",
			"UpdatedAt": "2023-11-05T21:59:50.080820909Z",
			"DeletedAt": null,
			"PlayerID": 2,
			"stateID": 20,
			"value": 0.42
		}
	],
	"level": {
		"ID": 0,
		"CreatedAt": "0001-01-01T00:00:00Z",
		"UpdatedAt": "0001-01-01T00:00:00Z",
		"DeletedAt": null,
		"PlayerID": 0,
		"levelID": 0
	}
}
```

Various fields are successfully updating (Active / Position / Rotation / Scale and States)

**Review by Balugrizzly — APPROVED** [review]
Changes Lookg good

**Comment by Balugrizzly** [comment]
@Sarahtonein Will merge this but feel free to make additional addition in a new branch

**Pull request merged**